lib/environment.sh

```diff
@@ -106,5 +106,9 @@ write_export() {
   if [ -w "$bp_dir" ]; then
     echo "export PATH=\"$build_dir/.heroku/node/bin:$build_dir/.heroku/yarn/bin:\$PATH:$build_dir/node_modules/.bin\"" > "$bp_dir/export"
     echo "export NODE_HOME=\"$build_dir/.heroku/node\"" >> "$bp_dir/export"
+    # only export this if the user hasn't configured a NODE_OPTIONS env variable
+    if [[ -z $NODE_OPTIONS ]]; then
```

Comment:
I think this conditional needs to live inside the `export` file itself, rather than here? Otherwise if an earlier/later buildpack does something with `NODE_OPTIONS` the conditional might evaluate differently if it runs now vs at time of the `export` file being sourced?

And in fact, instead of using a conditional, you could always use Bash's "default env var value if not set" syntax, like the Python buildpack uses:
https://github.com/heroku/heroku-buildpack-python/blob/5263022d8bb2f477fcbc1adace8d402e18285975/vendor/buildpack-stdlib_v8.sh#L76

Reply:
Is that the behavior we want here? Here's are the scenarios I'm thinking of:
- If the `NODE_OPTIONS` is a user set environment variable, **DO NOT** write the export
- If an earlier buildpack sets `NODE_OPTIONS` and this propagates to the Node.js buildpack, **DO NOT** write the export
- Node.js buildpack runs and no `NODE_OPTIONS` is set, **DO** write the export
- If a later buildpack sets `NODE_OPTIONS` that will only affect that buildpack unless they also export, right?

Or maybe I'm not getting how this `export` file is sourced later in the process.

Reply:
The export is re-run before every buildpack, and re-run in the order of the buildpacks.

So if there are three buildpacks, and all write an `export` file, the third buildpack's `bin/compile` effectively gets run like so:

```bash
source <buildpack_one>/export
source <buildpack_two>/export
<buildpack_three>/bin/compile
```

You are correct that in several scenarios it won't make any difference as to whether the conditional is inside the `export` file or not, but (a) there are presumably some where it would matter, (b) even if there weren't, if it's the same behaviour either way, then it feels more correct to have the conditional in the export file, like most other buildpacks do, and like all buildpacks do when setting default env vars in `.profile.d/` scripts (where it really does matter, since env vars can change between build time and run time).

Reply:
I made this change and tested it against an internal Ruby app:
- With no `NODE_OPTIONS` set, the default value was used
- With `NODE_OPTIONS` set, the given value was used